File: backend/src/ely_eye/compiler.py
from __future__ import annotations

import hashlib
from typing import Iterable

from .cache_fabric import CacheFabric
from .config import Settings, get_settings
from .db import Database
from .retrieval import RetrievalService
from .schemas import CompiledContext, ContextPlan, Modality, RetrievalHit, RuntimeProfile


PROFILE_TOKEN_BUDGETS = {
    RuntimeProfile.live_demo: 128_000,
    RuntimeProfile.extreme_context: 1_010_000,
    RuntimeProfile.library_100m: 262_144,
    RuntimeProfile.research_theater: 128_000,
}


class ContextCompiler:
    def __init__(self, settings: Settings | None = None, database: Database | None = None) -> None:
        self.settings = settings or get_settings()
        self.db = database or Database(self.settings)
        self.retrieval = RetrievalService(self.settings, self.db)
        self.cache = CacheFabric(self.settings, self.db)

    def compile(self, question: str, profile: RuntimeProfile) -> CompiledContext:
        cache_key = self.cache.semantic_key(question, profile.value)
        cached = self.cache.read_semantic_cache(cache_key)
        if cached:
            compiled = CompiledContext.model_validate(cached)
            if compiled.hits:
                return compiled

        plan = self.plan(question, profile)
        hits = self.retrieval.search(plan.retrieval_query, plan.evidence_budget_atoms)
        packed_text, token_total = self._pack_hits(hits, plan.token_budget, plan.retrieval_query)
        trace_id = hashlib.sha256(f"{cache_key}:{token_total}".encode("utf-8")).hexdigest()[:16]
        compiled = CompiledContext(
            plan=plan,
            hits=hits,
            packed_text=packed_text,
            token_equivalent=token_total,
            cache_trace_id=trace_id,
        )
        self.cache.write_semantic_cache(cache_key, compiled.model_dump(mode="json"))
        return compiled

    def plan(self, question: str, profile: RuntimeProfile) -> ContextPlan:
        modalities = infer_modalities(question)
        token_budget = PROFILE_TOKEN_BUDGETS[profile]
        evidence_budget = 48 if profile in {RuntimeProfile.library_100m, RuntimeProfile.extreme_context} else 24
        return ContextPlan(
            question=question,
            profile=profile,
            required_modalities=modalities,
            evidence_budget_atoms=evidence_budget,
            token_budget=token_budget,
            retrieval_query=question,
            compression_strategy=self._compression_strategy(profile, modalities),
            verifier_contract=[
                "Every factual claim must cite atom ids.",
                "Citations must include page, image region, frame, or code path when available.",
                "Contradictions must be reported as contradiction_notes.",
            ],
        )

    def _compression_strategy(self, profile: RuntimeProfile, modalities: Iterable[Modality]) -> str:
        has_visual = any(modality in {Modality.image, Modality.pdf_page, Modality.video_frame, Modality.ui_screenshot} for modality in modalities)
        if profile == RuntimeProfile.extreme_context:
            return "YaRN + V-NSA sparse blocks + DVKV visual KV compression + CPU/NVMe parked context"
        if profile == RuntimeProfile.library_100m:
            return "Context Cartridge retrieval + ColQwen-compatible dense vectors + BM25 + evidence graph pack"
        if has_visual:
            return "Visual patch cache + DVKV evidence-biased region retention"
        return "Prefix cache + sparse evidence packing"

    def _pack_hits(self, hits: list[RetrievalHit], token_budget: int, query: str) -> tuple[str, int]:
        sections: list[str] = []
        token_total = 0
        for hit in hits:
            atom = hit.atom
            evidence_text = focused_evidence_text(atom.text, query)
            atom_tokens = max(1, len(evidence_text) // 4)
            if token_total + atom_tokens > token_budget:
                break
            citation = {
                "atom_id": atom.atom_id,
                "modality": atom.modality.value,
                "source": atom.source,
                "image_ref": atom.image_ref,
                "layout": atom.layout.model_dump() if atom.layout else None,
                "score": hit.final_score,
            }
            sections.append(f"[EVIDENCE]\n{citation}\n{evidence_text}\n[/EVIDENCE]")
            token_total += atom_tokens
        return "\n\n".join(sections), token_total


def infer_modalities(question: str) -> list[Modality]:
    lowered = question.lower()
    modalities: list[Modality] = []
    if any(word in lowered for word in ("pdf", "page", "paper", "论文", "页面", "页码")):
        modalities.append(Modality.pdf_page)
    if any(word in lowered for word in ("image", "screenshot", "figma", "ui", "图", "截图", "设计")):
        modalities.extend([Modality.image, Modality.ui_screenshot])
    if any(word in lowered for word in ("video", "frame", "clip", "视频", "帧")):
        modalities.append(Modality.video_frame)
    if any(word in lowered for word in ("code", "function", "class", "代码", "函数")):
        modalities.append(Modality.code)
    if not modalities:
        modalities.append(Modality.mixed)
    return list(dict.fromkeys(modalities))


def focused_evidence_text(text: str, query: str, window: int = 1400, max_windows: int = 4) -> str:
    if len(text) <= window * max_windows:
        return text
    lowered = text.lower()
    query_terms = [term for term in query.lower().replace("/", " ").replace("-", " ").split() if len(term) > 2]
    anchors: list[int] = []
    for term in query_terms:
        index = lowered.find(term)
        if index >= 0:
            anchors.append(index)
    anchors.extend([0, len(text) // 3, (len(text) * 2) // 3])
    windows: list[tuple[int, int]] = []
    for anchor in anchors:
        start = max(0, anchor - window // 2)
        end = min(len(text), start + window)
        start = max(0, end - window)
        if all(abs(start - existing_start) > window // 2 for existing_start, _ in windows):
            windows.append((start, end))
        if len(windows) >= max_windows:
            break
    return "\n\n[...]\n\n".join(text[start:end].strip() for start, end in windows if text[start:end].strip())

File: ely-eye-mcp/internal/contextpack/compiler.go
package contextpack

import (
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"strings"
	"unicode/utf8"

	"github.com/ZacharyZhang-NY/ely-eye/ely-eye-mcp/internal/domain"
)

var profileBudgets = map[string]int64{
	"live_demo":        128000,
	"extreme_context":  1010000,
	"library_100m":     262144,
	"research_theater": 128000,
}

func Build(question string, profile string, hits []domain.AtomHit, requestedBudget int64) domain.EvidencePack {
	if profile == "" {
		profile = "library_100m"
	}
	budget := requestedBudget
	if budget <= 0 {
		budget = profileBudgets[profile]
	}
	if budget <= 0 {
		budget = profileBudgets["library_100m"]
	}
	var sections []string
	var tokens int64
	for _, hit := range hits {
		text := focusText(hit.Atom.Text, question, 1600)
		atomTokens := tokenEquivalent(text)
		if atomTokens == 0 {
			atomTokens = 1
		}
		if tokens+atomTokens > budget {
			break
		}
		sections = append(sections, evidenceSection(hit, text))
		tokens += atomTokens
	}
	trace := traceID(question, profile, tokens, hits)
	return domain.EvidencePack{
		Question:        question,
		Profile:         profile,
		TokenBudget:     budget,
		TokenEquivalent: tokens,
		CacheTraceID:    trace,
		Hits:            SummarizeHits(hits, 900),
		PackedText:      strings.Join(sections, "\n\n"),
		VerifierContract: []string{
			"Every factual claim must cite atom ids.",
			"Citations must include page, image region, frame, or code path when available.",
			"Contradictions must be reported as contradiction_notes.",
		},
	}
}

func SummarizeHits(hits []domain.AtomHit, maxRunes int) []domain.AtomSearchHit {
	summaries := make([]domain.AtomSearchHit, 0, len(hits))
	for _, hit := range hits {
		summaries = append(summaries, domain.AtomSearchHit{
			Atom:        summarizeAtom(hit.Atom, maxRunes),
			SparseScore: hit.SparseScore,
			FinalScore:  hit.FinalScore,
		})
	}
	return summaries
}

func summarizeAtom(atom domain.Atom, maxRunes int) domain.AtomSummary {
	return domain.AtomSummary{
		AtomID:          atom.AtomID,
		SourceID:        atom.SourceID,
		Modality:        atom.Modality,
		Source:          atom.Source,
		Time:            atom.Time,
		TextPreview:     trimRunes(atom.Text, maxRunes),
		ImageRef:        atom.ImageRef,
		TokenEquivalent: atom.TokenEquivalent,
	}
}

func evidenceSection(hit domain.AtomHit, text string) string {
	return fmt.Sprintf(
		"[EVIDENCE]\natom_id=%s\nmodality=%s\nsource=%s\nimage_ref=%s\nscore=%.4f\n%s\n[/EVIDENCE]",
		hit.Atom.AtomID,
		hit.Atom.Modality,
		hit.Atom.Source,
		hit.Atom.ImageRef,
		hit.FinalScore,
		text,
	)
}

func focusText(text string, query string, window int) string {
	if utf8.RuneCountInString(text) <= window {
		return text
	}
	lowerText := strings.ToLower(text)
	best := 0
	for _, term := range strings.Fields(strings.ToLower(query)) {
		if len(term) < 3 {
			continue
		}
		if index := strings.Index(lowerText, term); index >= 0 {
			best = index
			break
		}
	}
	start := best - window/2
	if start < 0 {
		start = 0
	}
	runes := []rune(text)
	runeStart := byteToRuneOffset(text, start)
	if runeStart+window > len(runes) {
		runeStart = len(runes) - window
	}
	if runeStart < 0 {
		runeStart = 0
	}
	return string(runes[runeStart : runeStart+window])
}

func trimRunes(text string, limit int) string {
	if limit <= 0 || utf8.RuneCountInString(text) <= limit {
		return text
	}
	runes := []rune(text)
	return string(runes[:limit])
}

func tokenEquivalent(text string) int64 {
	var ascii int64
	var nonASCII int64
	for _, r := range text {
		if r < 128 {
			ascii++
		} else {
			nonASCII++
		}
	}
	total := ascii/4 + nonASCII/2
	if total == 0 && text != "" {
		return 1
	}
	return total
}

func traceID(question string, profile string, tokens int64, hits []domain.AtomHit) string {
	hash := sha256.New()
	fmt.Fprintf(hash, "%s\n%s\n%d\n", question, profile, tokens)
	for _, hit := range hits {
		fmt.Fprintln(hash, hit.Atom.AtomID)
	}
	return hex.EncodeToString(hash.Sum(nil))[:16]
}

func byteToRuneOffset(text string, byteOffset int) int {
	if byteOffset <= 0 {
		return 0
	}
	count := 0
	for index := range text {
		if index >= byteOffset {
			return count
		}
		count++
	}
	return count
}
